File: api/groups.php
<?php
/**
 * API: Groups
 * 
 * POST /create - Create group
 * POST /join - Join by code
 * GET  /feed - Group feed (published reports only!)
 * GET  /my - My group info
 */

require_once __DIR__ . '/../config/database.php';

header('Content-Type: application/json');
header('Access-Control-Allow-Origin: https://skilltracer.art-artel.su');
header('Access-Control-Allow-Headers: X-Init-Data, Content-Type');
header('Access-Control-Allow-Methods: GET, POST, OPTIONS');

if ($_SERVER['REQUEST_METHOD'] === 'OPTIONS') {
    http_response_code(200);
    exit;
}

$botToken = '7973502371:AAGZ1A5XeWdKaiMKZDumfTa9gCr0I3a8EMg';
$initData = $_SERVER['HTTP_X_INIT_DATA'] ?? '';

if (empty($initData)) {
    http_response_code(401);
    echo json_encode(['error' => 'Missing init data']);
    exit;
}

parse_str($initData, $data);
$userData = json_decode($data['user'] ?? '{}', true);
$userId = $userData['id'] ?? null;

if (!$userId) {
    http_response_code(401);
    echo json_encode(['error' => 'Invalid user']);
    exit;
}

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';

try {
    $db = getDB();
    
    // Ensure user exists
    $stmt = $db->prepare("SELECT id FROM users WHERE id = ?");
    $stmt->execute([$userId]);
    if (!$stmt->fetch()) {
        $stmt = $db->prepare("
            INSERT INTO users (id, username, first_name, last_name)
            VALUES (?, ?, ?, ?)
        ");
        $stmt->execute([
            $userId,
            $userData['username'] ?? null,
            $userData['first_name'] ?? null,
            $userData['last_name'] ?? null,
        ]);
    }
    
    if ($method === 'POST' && $action === 'create') {
        // Check if already in group
        $stmt = $db->prepare("SELECT group_id FROM group_members WHERE user_id = ?");
        $stmt->execute([$userId]);
        if ($stmt->fetch()) {
            http_response_code(400);
            echo json_encode(['error' => 'Already in a group']);
            exit;
        }
        
        $input = json_decode(file_get_contents('php://input'), true);
        $name = $input['name'] ?? 'My Group';
        $description = $input['description'] ?? null;
        
        // Generate invite code
        $code = strtoupper(substr(str_shuffle('ABCDEFGHJKMNPQRSTUVWXYZ23456789'), 0, 8));
        
        // Create group
        $stmt = $db->prepare("
            INSERT INTO groups (name, invite_code, owner_id, description)
            VALUES (?, ?, ?, ?)
        ");
        $stmt->execute([$name, $code, $userId, $description]);
        $groupId = $db->lastInsertId();
        
        // Add owner as member
        $stmt = $db->prepare("
            INSERT INTO group_members (group_id, user_id, role)
            VALUES (?, ?, 'owner')
        ");
        $stmt->execute([$groupId, $userId]);
        
        echo json_encode([
            'id' => $groupId,
            'name' => $name,
            'invite_code' => $code,
            'member_count' => 1,
        ]);
        
    } elseif ($method === 'POST' && $action === 'join') {
        $input = json_decode(file_get_contents('php://input'), true);
        $code = strtoupper($input['invite_code'] ?? '');
        
        // Check if already in group
        $stmt = $db->prepare("SELECT group_id FROM group_members WHERE user_id = ?");
        $stmt->execute([$userId]);
        if ($stmt->fetch()) {
            http_response_code(400);
            echo json_encode(['error' => 'Already in a group']);
            exit;
        }
        
        // Find group
        $stmt = $db->prepare("SELECT id FROM groups WHERE invite_code = ?");
        $stmt->execute([$code]);
        $group = $stmt->fetch();
        
        if (!$group) {
            http_response_code(404);
            echo json_encode(['error' => 'Group not found']);
            exit;
        }
        
        $groupId = $group['id'];
        
        // Check member limit (max 3)
        $stmt = $db->prepare("SELECT COUNT(*) FROM group_members WHERE group_id = ?");
        $stmt->execute([$groupId]);
        $count = $stmt->fetchColumn();
        
        if ($count >= 3) {
            http_response_code(400);
            echo json_encode(['error' => 'Group is full (max 3 members)']);
            exit;
        }
        
        // Add member
        $stmt = $db->prepare("
            INSERT INTO group_members (group_id, user_id, role)
            VALUES (?, ?, 'member')
        ");
        $stmt->execute([$groupId, $userId]);
        
        echo json_encode(['status' => 'joined', 'group_id' => $groupId]);
        
    } elseif ($method === 'GET' && $action === 'my') {
        // Get my group
        $stmt = $db->prepare("
            SELECT g.*, COUNT(gm.user_id) as member_count
            FROM groups g
            JOIN group_members gm ON g.id = gm.group_id
            WHERE g.id IN (SELECT group_id FROM group_members WHERE user_id = ?)
            GROUP BY g.id
        ");
        $stmt->execute([$userId]);
        $group = $stmt->fetch();
        
        if (!$group) {
            http_response_code(404);
            echo json_encode(['error' => 'Not in a group']);
            exit;
        }
        
        echo json_encode([
            'id' => $group['id'],
            'name' => $group['name'],
            'invite_code' => $group['invite_code'],
            'description' => $group['description'],
            'member_count' => (int)$group['member_count'],
        ]);
        
    } elseif ($method === 'GET' && $action === 'feed') {
        // Get feed (published reports only!)
        $limit = min((int)($_GET['limit'] ?? 10), 50);
        $offset = (int)($_GET['offset'] ?? 0);
        
        // Find user's group
        $stmt = $db->prepare("SELECT group_id FROM group_members WHERE user_id = ?");
        $stmt->execute([$userId]);
        $membership = $stmt->fetch();
        
        if (!$membership) {
            http_response_code(404);
            echo json_encode(['error' => 'Not in a group']);
            exit;
        }
        
        $groupId = $membership['group_id'];
        
        // Get group member IDs (exclude self)
        $stmt = $db->prepare("
            SELECT user_id FROM group_members 
            WHERE group_id = ? AND user_id != ?
        ");
        $stmt->execute([$groupId, $userId]);
        $memberIds = $stmt->fetchAll(PDO::FETCH_COLUMN);
        
        if (empty($memberIds)) {
            echo json_encode([]);
            exit;
        }
        
        // Get published reports from group members
        $placeholders = implode(',', array_fill(0, count($memberIds), '?'));
        $sql = "
            SELECT wr.*, u.id as user_id, u.first_name, u.photo_url
            FROM week_reports wr
            JOIN users u ON wr.user_id = u.id
            WHERE wr.user_id IN ($placeholders)
            AND wr.status = 'published'
            ORDER BY wr.published_at DESC
            LIMIT ? OFFSET ?
        ";
        
        $params = array_merge($memberIds, [$limit, $offset]);
        $stmt = $db->prepare($sql);
        $stmt->execute($params);
        $reports = $stmt->fetchAll();
        
        $result = [];
        foreach ($reports as $r) {
            $result[] = [
                'id' => $r['id'],
                'user' => [
                    'id' => $r['user_id'],
                    'first_name' => $r['first_name'],
                    'photo_url' => $r['photo_url'],
                ],
                'week_start' => $r['week_start_date'],
                'week_end' => $r['week_end_date'],
                'avg_mood' => $r['avg_mood'] ? (float)$r['avg_mood'] : null,
                'filled_days' => (int)$r['filled_days'],
                'metrics_summary' => json_decode($r['metrics_summary'] ?? '{}', true),
                'published_at' => $r['published_at'],
            ];
        }
        
        echo json_encode($result);
        
    } else {
        http_response_code(400);
        echo json_encode(['error' => 'Invalid action']);
    }
    
} catch (Exception $e) {
    http_response_code(500);
    echo json_encode(['error' => 'Database error: ' . $e->getMessage()]);
}

File: backend/app/api/v1/endpoints/groups.py
"""
Groups API

Управление группами и групповой лентой.
"""

import logging
from datetime import date, timedelta
from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func
from pydantic import BaseModel, Field

from app.api.deps import get_current_user, get_db
from app.models import User, Group, GroupMember, WeekReport, ReportStatus, GroupRole

logger = logging.getLogger(__name__)
router = APIRouter()


# =============================================================================
# Schemas
# =============================================================================

class GroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = None


class JoinGroup(BaseModel):
    invite_code: str = Field(..., min_length=6, max_length=32)


class GroupOut(BaseModel):
    id: int
    name: str
    invite_code: str
    description: Optional[str]
    member_count: int
    
    class Config:
        from_attributes = True


class GroupMemberOut(BaseModel):
    user_id: int
    first_name: str
    photo_url: Optional[str]
    role: str


class FeedItemOut(BaseModel):
    id: int
    user: dict
    week_start: date
    week_end: date
    avg_mood: Optional[float]
    filled_days: int
    metrics_summary: dict
    published_at: date
    
    class Config:
        from_attributes = True


# =============================================================================
# Endpoints
# =============================================================================

@router.post("", response_model=GroupOut)
async def create_group(
    data: GroupCreate,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Создание группы."""
    # Проверяем что пользователь не в группе
    result = await db.execute(
        select(GroupMember).where(GroupMember.user_id == user.id)
    )
    if result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Already in a group",
        )
    
    # Создаем группу
    group = Group(
        name=data.name,
        invite_code=Group.generate_invite_code(),
        owner_id=user.id,
        description=data.description,
    )
    db.add(group)
    await db.flush()
    
    # Добавляем создателя как owner
    member = GroupMember(
        group_id=group.id,
        user_id=user.id,
        role=GroupRole.OWNER,
    )
    db.add(member)
    await db.commit()
    await db.refresh(group)
    
    return GroupOut(
        id=group.id,
        name=group.name,
        invite_code=group.invite_code,
        description=group.description,
        member_count=1,
    )


@router.post("/join")
async def join_group(
    data: JoinGroup,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Присоединение к группе по коду."""
    # Проверяем что пользователь не в группе
    result = await db.execute(
        select(GroupMember).where(GroupMember.user_id == user.id)
    )
    if result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Already in a group",
        )
    
    # Ищем группу
    result = await db.execute(
        select(Group).where(Group.invite_code == data.invite_code.upper())
    )
    group = result.scalar_one_or_none()
    
    if not group:
        raise HTTPException(status_code=404, detail="Group not found")
    
    # Проверяем лимит (3 человека)
    result = await db.execute(
        select(func.count(GroupMember.user_id)).where(GroupMember.group_id == group.id)
    )
    member_count = result.scalar()
    
    if member_count >= 3:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Group is full (max 3 members)",
        )
    
    # Добавляем
    member = GroupMember(
        group_id=group.id,
        user_id=user.id,
        role=GroupRole.MEMBER,
    )
    db.add(member)
    await db.commit()
    
    return {"status": "joined", "group_id": group.id}


@router.get("/my", response_model=GroupOut)
async def get_my_group(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Моя группа."""
    result = await db.execute(
        select(GroupMember).where(GroupMember.user_id == user.id)
    )
    membership = result.scalar_one_or_none()
    
    if not membership:
        raise HTTPException(status_code=404, detail="Not in a group")
    
    result = await db.execute(
        select(Group).where(Group.id == membership.group_id)
    )
    group = result.scalar_one()
    
    # Считаем участников
    result = await db.execute(
        select(func.count(GroupMember.user_id)).where(GroupMember.group_id == group.id)
    )
    member_count = result.scalar()
    
    return GroupOut(
        id=group.id,
        name=group.name,
        invite_code=group.invite_code,
        description=group.description,
        member_count=member_count,
    )


@router.get("/members", response_model=List[GroupMemberOut])
async def get_group_members(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """Участники моей группы."""
    # Находим группу пользователя
    result = await db.execute(
        select(GroupMember).where(GroupMember.user_id == user.id)
    )
    membership = result.scalar_one_or_none()
    
    if not membership:
        raise HTTPException(status_code=404, detail="Not in a group")
    
    # Получаем участников
    result = await db.execute(
        select(GroupMember, User)
        .join(User, GroupMember.user_id == User.id)
        .where(GroupMember.group_id == membership.group_id)
    )
    members = result.all()
    
    return [
        GroupMemberOut(
            user_id=u.id,
            first_name=u.first_name or "User",
            photo_url=u.photo_url,
            role=m.role.value,
        )
        for m, u in members
    ]


@router.get("/feed", response_model=List[FeedItemOut])
async def get_group_feed(
    limit: int = 10,
    offset: int = 0,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user),
):
    """
    Групповая лента (только published отчеты).
    Ключевая фича - видны только опубликованные отчеты!
    """
    # Находим группу
    result = await db.execute(
        select(GroupMember).where(GroupMember.user_id == user.id)
    )
    membership = result.scalar_one_or_none()
    
    if not membership:
        raise HTTPException(status_code=404, detail="Not in a group")
    
    # Получаем ID участников группы
    result = await db.execute(
        select(GroupMember.user_id).where(
            GroupMember.group_id == membership.group_id,
            GroupMember.user_id != user.id,  # Не показываем свои отчеты в ленте
        )
    )
    member_ids = [row[0] for row in result.all()]
    
    if not member_ids:
        return []
    
    # Получаем published отчеты
    result = await db.execute(
        select(WeekReport, User)
        .join(User, WeekReport.user_id == User.id)
        .where(
            WeekReport.user_id.in_(member_ids),
            WeekReport.status == ReportStatus.PUBLISHED,
        )
        .order_by(WeekReport.published_at.desc())
        .limit(limit)
        .offset(offset)
    )
    reports = result.all()
    
    return [
        FeedItemOut(
            id=r.id,
            user={
                "id": u.id,
                "first_name": u.first_name,
                "photo_url": u.photo_url,
            },
            week_start=r.week_start_date,
            week_end=r.week_end_date,
            avg_mood=r.avg_mood,
            filled_days=r.filled_days,
            metrics_summary=r.metrics_summary,
            published_at=r.published_at.date() if r.published_at else r.week_start_date,
        )
        for r, u in reports
    ]
